### PR TITLE
SapMachine #1646: Provide a mechanism to augment the Http KeepAliveCache key

### LINKED ISSUE #1646: Provide a mechanism to augment the Http KeepAliveCache key
A customer is accessing an HttpServer backend via HttpURLConnection through JAX-WS. Connections get authenticated by some special mechanism on a per-socket base and subsequent HTTP requests need to happen on the authenticated connection. So the matching authenticated connection has to be retrieved from the cache. To achieve this, we add a field of type ThreadLocal<String> to sun.net.www.http.KeepAliveCache that can be armed with a value and serves as an additional part of the KeepAliveCache key.

We have thought about alternative solutions for the customer but at the moment there seem to be none. The feature does not affect the API and is only enabled when system property `com.sap.jvm.UseHttpKeepAliveCacheKeyExtension` is set to `true`. To use this feature, client code needs an export statement like `--add-exports java.base/sun.net.www.http=ALL-UNNAMED`.

### PR DESCRIPTION
**Pull request by RealCLanger**
Adds a mechanism to hint the HttpClient's keepAliveCache, requested  by customer.

fixes #1646

### CODE REVIEW

**Comment by SapMachine** [comment]
Hello @RealCLanger, this pull request fulfills all formal requirements.

**Comment by SapMachine** [comment]
Hello @RealCLanger, this pull request fulfills all formal requirements.

**Comment by SapMachine** [comment]
Hello @RealCLanger, this pull request fulfills all formal requirements.

**Comment by schmelter-sap** [comment]
The copyright headers have to be updated to 2024.

File: src/java.base/share/classes/sun/net/www/http/KeepAliveCache.java [hunk 1]
```diff
@@ -378,10 +386,26 @@ class KeepAliveKey {
...
      */
     public KeepAliveKey(URL url, Object obj) {
+        // SapMachine 2024-04-12: Provide additional key field for KeepAliveCache entries
+        final record KeyObject(String ci, Object obj) {
+            @Override
+            public boolean equals(Object other) {
+                if (this == other) {
+                    return true;
+                }
+                if (other == null || !(other instanceof KeyObject)) {
```
Comment (schmelter-sap) [inline src/java.base/share/classes/sun/net/www/http/KeepAliveCache.java]: Shouldn't instanceof handle null correctly?

File: src/java.base/share/classes/sun/net/www/http/KeepAliveCache.java [hunk 2]
```diff
@@ -378,10 +386,26 @@ class KeepAliveKey {
      * @param url the URL containing the protocol, host and port information
      */
     public KeepAliveKey(URL url, Object obj) {
+        // SapMachine 2024-04-12: Provide additional key field for KeepAliveCache entries
+        final record KeyObject(String ci, Object obj) {
```
Comment (schmelter-sap) [inline src/java.base/share/classes/sun/net/www/http/KeepAliveCache.java]: I assume 'ci' stands for connection id?

**Comment by RealCLanger** [comment]
> The copyright headers have to be updated to 2024.

To keep the diff low, we usually don't update the copyright years in files that come from upstream.

File: src/java.base/share/classes/sun/net/www/http/KeepAliveCache.java [hunk 3]
```diff
@@ -392,10 +416,11 @@ public boolean equals(Object obj) {
         if ((obj instanceof KeepAliveKey) == false)
             return false;
         KeepAliveKey kae = (KeepAliveKey)obj;
+        // SapMachine 2024-04-12: Provide additional key field for KeepAliveCache entries
```
Comment (TheRealMDoerr) [inline src/java.base/share/classes/sun/net/www/http/KeepAliveCache.java]: Maybe move comment down to the modified code?

Comment (RealCLanger) [reply src/java.base/share/classes/sun/net/www/http/KeepAliveCache.java]: Done.

**Comment by SapMachine** [comment]
Hello @RealCLanger, this pull request fulfills all formal requirements.

**Review by TheRealMDoerr — APPROVED** [review]
LGTM.

**Review by schmelter-sap — APPROVED** [review]
Looks good

File: src/java.base/share/classes/sun/net/www/http/KeepAliveCache.java (hunk 2 above)
Comment (RealCLanger) [reply src/java.base/share/classes/sun/net/www/http/KeepAliveCache.java]: Yes, I made it more explicit.

**Comment by SapMachine** [comment]
Hello @RealCLanger, this pull request fulfills all formal requirements.

File: src/java.base/share/classes/sun/net/www/http/KeepAliveCache.java (hunk 1 above)
Comment (RealCLanger) [reply src/java.base/share/classes/sun/net/www/http/KeepAliveCache.java]: You're right. I improved the coding a bit. Thanks for the finding!

**Review by TheRealMDoerr — APPROVED** [review]
Still good.

**Comment by SapMachine** [comment]
Hello @RealCLanger, this pull request fulfills all formal requirements.

**Pull request merged**